src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+                    return;
+                case 4:
+                    // yield return 43;
+                    yieldReturn(43, state: 5);
+                    return;
+                case 5:
+                    // yield return 44;
+                    yieldReturn(44, state: 6);
+                    return;
+                case 6:
+                    // await Done();
+                    awaitExpr(state: 7, fast: true);
+                    goto case 7;
+                case 7:
+                    // yield return 45;
+                    yieldReturn(45, state: 9);
+                    return;
+                case 9:
+                    // yield return 46;
+                    yieldReturn(46, state: 10);
+                    return;
+                case 10:
+                    end();
+                    return;
+            }
+
+            // yield handshake:
+            void yieldReturn(int value, int state)
+            {
+                _current = value;
+
+                int previousState = State;
+                State = state;
+
+                if (this._promiseIsActive)
+                {
+                    // reaching a `yield` following an `await` or directly from start
+                    _valueOrEndPromise.SetResult(true);
+                }
+            }
+
+            // await handshake:
+            void awaitExpr(int state, bool fast)
+            {
+                Task task = fast ? Task.CompletedTask : Task.Delay(100);
+                this.Awaiter = task.GetAwaiter();
+                State = state;
+                if (!task.IsCompleted)
+                {
+                    if (!this._promiseIsActive)
+                    {
+                        _promiseIsActive = true;
+                        _valueOrEndPromise.Reset();
+                    }
+                    var self = this;
+                    Builder.AwaitOnCompleted(ref Awaiter, ref self);
+                }
+            }
+
+            void end()
+            {
+                State = StateMachineStates.FinishedStateMachine;
+                if (!_promiseIsActive)
+                {
+                    _promiseIsActive = true;
+                    _valueOrEndPromise.Reset();
+                }
+                _valueOrEndPromise.SetResult(false);
+            }
+        }
+
+        // Contract: WaitForNextAsync will either return a false (no value left) or a true (found a value) with the promise set (to signal an unreturned value is stored).
+        public ValueTask<bool> WaitForNextAsync()
+        {
+            // if we have a pending promise already (ie. the async code was started by last TryGetNext), then we just return it 
+            // if we don't have a pending promise, then call MoveNext()
+            if (!this._promiseIsActive || State == StateMachineStates.NotStartedStateMachine)
+            {
+                MoveNext();
```

Comment:
I think we probably also need to use Builder.Start to invoke MoveNext, similar to how I did in my prototype:
https://github.com/dotnet/corefx/blob/ecf20bcd7ce610c6cd242441c44af4389295757d/src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSourceTests.cs#L170
That will ensure that any changes to ExecutionContext inside of the async iterator don't leak out to the caller (unless we decide we don't care about that).